### DISCUSSION TITLE
brennenstuhl®Connect Zigbee thermostat HT CZ 01

### DISCUSSION THREAD

**Discussion by csidirop** [Koenkk/zigbee2mqtt#14080] [Devices]
Hi everyone, I need help setting up the [brennenstuhl thermostat](https://www.brennenstuhl.com/en-DE/products/smart-home/brennenstuhl-r-connect-zigbee-radiator-thermostat-ht-cz-01).

I followed the instructions to [support new devices](https://www.zigbee2mqtt.io/advanced/support-new-devices/01_support_new_devices.html#instructions) and created an external converter file:

```const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;

const definition = {
    //zigbeeModel: ['TS0601'], // The model ID from: Device with modelID 'lumi.sens' is not supported.
    fingerprint: [
        {
            // The model ID from: Device with modelID 'TS0601' is not supported
            // You may need to add \u0000 at the end of the name in some cases
            modelID: 'TS0601',
            // The manufacturer name from: Device with modelID 'TS0601' is not supported.
            manufacturerName: '_TZE200_2cs6g9i7'
        },
    ],
    model: 'HTCZ01', // Vendor model number, look on the device for a model number
    vendor: 'brennenstuhl', // Vendor of the device (only used for documentation and startup logging)
    description: 'Smart Home Heizkörperthermostat', // Description
    fromZigbee: [
        fz.ignore_basic_report, // Add this if you are getting no converter for 'genBasic'
        fz.tuya_data_point_dump, // This is a debug converter, it will be described in the next part
        fz.temperature
    ], 
    toZigbee: [
        tz.tuya_data_point_test, // Another debug converter
    ],
    configure: async (device, coordinatorEndpoint, logger) => {
        const endpoint = device.getEndpoint(1);
        await reporting.bind(endpoint, coordinatorEndpoint, ['genBasic']);
    },
    exposes: [ // Defines what this device exposes, used for e.g. Home Assistant discovery and in the frontend
        e.battery(),
        e.temperature(),
        e.humidity()], 
};

module.exports = definition;
```   
At this point the device pairs, fills the log with messages and I don't know how to proceed. So I looked for other thermostats in [zigbee-herdsman-converter](https://github.com/Koenkk/zigbee-herdsman-converters/tree/master/devices) and found following: As you can see, the device reports with Zigbee model 'TS0601' and manufacturer name '_TZE200_2cs6g9i7'. So it might be a [TuYa TS0601 thermostat](https://www.zigbee2mqtt.io/devices/TS0601_thermostat.html) inside. 
I would like to test if it responds correctly when set up as one of those. It thought adding it here https://github.com/Koenkk/zigbee-herdsman-converters/blob/master/devices/tuya.js#L1856 should be enough to test it.

Any opinions, corrections or hints?

**Comment by Koenkk** [comment]
> So it might be a [TuYa TS0601 thermostat](https://www.zigbee2mqtt.io/devices/TS0601_thermostat.html) inside.

Try with:

```js
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const ota = require('zigbee-herdsman-converters/lib/ota');
const tuya = require('zigbee-herdsman-converters/lib/tuya');
const utils = require('zigbee-herdsman-converters/lib/utils');
const e = exposes.presets;
const ea = exposes.access;

const definition = {
    zigbeeModel: ['kud7u2l'],
    fingerprint: [
        {modelID: 'TS0601', manufacturerName: '_TZE200_ckud7u2l'},
        {modelID: 'TS0601', manufacturerName: '_TZE200_ywdxldoj'},
        {modelID: 'TS0601', manufacturerName: '_TZE200_cwnjrr72'},
        {modelID: 'TS0601', manufacturerName: '_TZE200_pvvbommb'},
        {modelID: 'TS0601', manufacturerName: '_TZE200_9sfg7gm0'}, // HomeCloud
        {modelID: 'TS0601', manufacturerName: '_TZE200_2atgpdho'}, // HY367
        {modelID: 'TS0601', manufacturerName: '_TZE200_cpmgn2cf'},
        {modelID: 'TS0601', manufacturerName: '_TZE200_4eeyebrt'}, // Immax 07732B
        {modelID: 'TS0601', manufacturerName: '_TZE200_8whxpsiw'}, // EVOLVEO
        {modelID: 'TS0601', manufacturerName: '_TZE200_2cs6g9i7'}, // brennenstuhl
    ],
    model: 'TS0601_thermostat',
    vendor: 'TuYa',
    description: 'Radiator valve with thermostat',
    whiteLabel: [
        {vendor: 'Moes', model: 'HY368'},
        {vendor: 'Moes', model: 'HY369RT'},
        {vendor: 'SHOJZJ', model: '378RT'},
        {vendor: 'Silvercrest', model: 'TVR01'},
        {vendor: 'Immax', model: '07732B'},
    ],
    meta: {tuyaThermostatPreset: tuya.thermostatPresets, tuyaThermostatSystemMode: tuya.thermostatSystemModes3},
    ota: ota.zigbeeOTA,
    onEvent: tuya.onEventSetLocalTime,
    fromZigbee: [fz.tuya_thermostat, fz.ignore_basic_report, fz.ignore_tuya_set_time],
    toZigbee: [tz.tuya_thermostat_child_lock, tz.tuya_thermostat_window_detection, tz.tuya_thermostat_valve_detection,
        tz.tuya_thermostat_current_heating_setpoint, tz.tuya_thermostat_auto_lock,
        tz.tuya_thermostat_calibration, tz.tuya_thermostat_min_temp, tz.tuya_thermostat_max_temp,
        tz.tuya_thermostat_boost_time, tz.tuya_thermostat_comfort_temp, tz.tuya_thermostat_eco_temp,
        tz.tuya_thermostat_force_to_mode, tz.tuya_thermostat_force, tz.tuya_thermostat_preset, tz.tuya_thermostat_away_mode,
        tz.tuya_thermostat_window_detect, tz.tuya_thermostat_schedule, tz.tuya_thermostat_week, tz.tuya_thermostat_away_preset,
        tz.tuya_thermostat_schedule_programming_mode],
    exposes: [
        e.child_lock(), e.window_detection(),
        exposes.binary('window_open', ea.STATE).withDescription('Window open?'),
        e.battery_low(), e.valve_detection(), e.position(),
        exposes.climate().withSetpoint('current_heating_setpoint', 5, 35, 0.5, ea.STATE_SET)
            .withLocalTemperature(ea.STATE).withSystemMode(['heat', 'auto', 'off'], ea.STATE_SET,
                'Mode of this device, in the `heat` mode the TS0601 will remain continuously heating, i.e. it does not regulate ' +
                'to the desired temperature. If you want TRV to properly regulate the temperature you need to use mode `auto` ' +
                'instead setting the desired temperature.')
            .withLocalTemperatureCalibration(-9, 9, 1, ea.STATE_SET)
            .withPreset(['schedule', 'manual', 'boost', 'complex', 'comfort', 'eco'])
            .withRunningState(['idle', 'heat'], ea.STATE),
        e.auto_lock(), e.away_mode(), e.away_preset_days(), e.boost_time(), e.comfort_temperature(), e.eco_temperature(), e.force(),
        e.max_temperature(), e.min_temperature(), e.away_preset_temperature(),
        exposes.composite('programming_mode').withDescription('Schedule MODE ⏱ - In this mode, ' +
                'the device executes a preset week programming temperature time and temperature.')
            .withFeature(e.week())
            .withFeature(exposes.text('workdays_schedule', ea.STATE_SET))
            .withFeature(exposes.text('holidays_schedule', ea.STATE_SET))],
};

module.exports = definition;
```

**Reply by csidirop** [reply]
That worked well but some properties showing null or not working at all:
![grafik](https://user-images.githubusercontent.com/43964592/191071171-8397963a-4dfd-4a33-ace9-10c1cca8e26e.png)

What can I do? Does that mean that this isn't a TuYa device or that brennenstuhl modified it?  Or am i missing something?

Thanks!

**Reply by Koenkk** [reply]
That means it probably uses different data points or doesn't support these features at all. To find this out you have to check it with the original TuYa gateway. 

https://www.zigbee2mqtt.io/advanced/support-new-devices/03_find_tuya_data_points.html#requirements-and-caveats

**Reply by csidirop** [reply]
Is this the only solution?

**Reply by Koenkk** [reply]
Yes, unfortunately TuYa has the bad habit to change their API for each device :(

**Reply by csidirop** [reply]
Okay, I'll give it a try. I found this gateway here: https://www.aliexpress.com/item/1005002762399241.html Will this work? It seems like there are a lot whitelabeled TuYa gateways out there...

**Reply by Koenkk** [reply]
That one should do it.

**Reply by csidirop** [reply]
Okay, that looked like the "most TuYa" gateway I found and it was cheap 😅 lets hope it will arrive. I'll come back in 2-3 weeks if I have issues with the retrieval of the data points. Thanks!

**Reply by csidirop** [reply]
Hi again, I got all data points: 

```
Window state:		17
Set Temperature:	2
Window check:		18
Current temperature: 	3
Valve State:		19
Mode:			4
Valve testing:		20
BAttery capacity:	21
TempSet:		102
Child Lock:		7
Temperature correction:	27
```
and have some log messages referencing the dps:

```
2022-10-05 17:37:12Received Zigbee message from '0x84fd27fffe0cf606', type 'commandDataResponse', cluster 'manuSpecificTuya', data '{"dpValues":[{"data":{"data":[0],"type":"Buffer"},"datatype":4,"dp":17}],"seq":51969}' from endpoint 1 with groupID 0
2022-10-05 17:37:12Received Zigbee message from '0x84fd27fffe0cf606', type 'commandDataResponse', cluster 'manuSpecificTuya', data '{"dpValues":[{"data":{"data":[0],"type":"Buffer"},"datatype":4,"dp":19}],"seq":52225}' from endpoint 1 with groupID 0
2022-10-05 17:37:12Received Zigbee message from '0x84fd27fffe0cf606', type 'commandDataResponse', cluster 'manuSpecificTuya', data '{"dpValues":[{"data":{"data":[0,0,0,237],"type":"Buffer"},"datatype":2,"dp":3}],"seq":52481}' from endpoint 1 with groupID 0
2022-10-05 17:37:12Received Zigbee message from '0x84fd27fffe0cf606', type 'commandDataResponse', cluster 'manuSpecificTuya', data '{"dpValues":[{"data":{"data":[0,0,0,100],"type":"Buffer"},"datatype":2,"dp":21}],"seq":52737}' from endpoint 1 with groupID 0
```

and tried [adding ](https://www.zigbee2mqtt.io/advanced/support-new-devices/02_support_new_tuya_devices.html#_5-adding-your-first-data-point)it but I'm a bit lost again. I tired a bit with your example but in the end I have no idea have to do. Can you help me out again?

**Reply by Koenkk** [reply]
I'm working on a new api to integrate tuya devices. I've created a first version for this device: https://gist.github.com/Koenkk/116d770830217a1bca3ccbd456c6e94a

Can you check if it works and extend if possible? Make sure you are using the latest z2m dev. (https://www.zigbee2mqtt.io/advanced/more/switch-to-dev-branch.html)

**Reply by csidirop** [reply]
Great news! Yes, I'll try tomorrow.

**Reply by csidirop** [reply]
So I changed to the dev branch and tried your suggestion. As expected, the four properties were shown (even if without values): 
![grafik](https://user-images.githubusercontent.com/43964592/194588619-b5fc71f8-5c19-4704-98cd-b6651e7db415.png)
Trying to set them resulted in errors, eg.: `No converter available for 'current_heating_setpoint' (8)`

**Reply by Koenkk** [reply]
whoops; updated https://gist.github.com/Koenkk/116d770830217a1bca3ccbd456c6e94a

**Reply by dreesi** [reply]
Got one also and with your addition I can now work with it in home assistant, thank you very much :) looking forward to see it in the stable branch

**Reply by csidirop** [reply]
Sorry for the late response, I currently have very little time. But I tried adding new data points but with little success. I maybe need to double check the data points, something looked odd. I try as soon as I can and give feedback.

Btw. bought two additional, so simply out of self interest I want them to work as soon as possible :D

**Reply by marcelebbrecht** [reply]
Hi,

maybe this is a good starting-point to get more info about additional datapoints?

https://github.com/zigpy/zha-device-handlers/blob/0ad9fa2f6fd2cd53c9c010553d46dbc8092c1a48/zhaquirks/tuya/ts0601_trv.py#L1466-L1475

**Reply by djmaze** [reply]
Would be really interested to know especially if it sends the _Valve State_ correctly. Having a hard time finding a TRV which supports that.

**Reply by marcelebbrecht** [reply]
Finally made some progress:

```javascript
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const ota = require('zigbee-herdsman-converters/lib/ota');
const tuya = require('zigbee-herdsman-converters/lib/tuya');
const utils = require('zigbee-herdsman-converters/lib/utils');
const e = exposes.presets;
const ea = exposes.access;

const tzDataPoints = {
    ...tuya.tzDataPoints,
    key: ['current_heating_setpoint', 'local_temperature', 'window_detection'],
};

const definition = {
    fingerprint: tuya.fingerprint('TS0601', ['_TZE200_2cs6g9i7']),
    model: 'TS0601_thermostat',
    vendor: 'TuYa',
    description: 'Radiator valve with thermostat',
    // Whitelabel seems to have no effect?
    // How to achive to get the model HTCZ01 and vendor Brennenstuhl and an icon (this may be the icon of the TS0601_thermostat) to the webinterface
    whiteLabel: [
        {vendor: 'HTCZ01', model: 'Brennenstuhl'},
    ],
    description: 'Radiator valve with thermostat',
    // fromZigbee: [tuya.fzDataPoints],
    // Enable for recording unknown datapoints
    fromZigbee: [tuya.fzDataPoints, fz.tuya_data_point_dump],
    // Disabled as write failes:
    // 2022-11-01 19:57:38: Error: Write 0x84fd27fffe8af1b8/1 hvacThermostat({"systemMode":1}, {"sendWhen":"immediate","timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null,"writeUndiv":false}) failed (Status 'UNSUPPORTED_ATTRIBUTE')
    // at Endpoint.checkStatus (/srv/zigbee2mqtt/zigbee2mqtt/node_modules/zigbee-herdsman/src/controller/model/endpoint.ts:317:28)
    // at Endpoint.write (/srv/zigbee2mqtt/zigbee2mqtt/node_modules/zigbee-herdsman/src/controller/model/endpoint.ts:397:22)
    // at Object.convertSet (/srv/zigbee2mqtt/zigbee2mqtt/node_modules/zigbee-herdsman-converters/converters/toZigbee.js:1199:13)
    // at Publish.onMQTTMessage (/srv/zigbee2mqtt/zigbee2mqtt/lib/extension/publish.ts:246:36)
    // toZigbee: [tzDataPoints, tz.thermostat_system_mode],
    toZigbee: [tzDataPoints, tz.tuya_thermostat_child_lock],
    configure: tuya.configureMagicPacket,
    onEvent: tuya.onEventSetLocalTime,
    exposes: [
        exposes.climate()
            .withSetpoint('current_heating_setpoint', 5, 35, 0.5, ea.STATE_SET)
            .withLocalTemperature(ea.STATE).withSystemMode(['off', 'auto', 'heat'], ea.STATE),
            // how to get 'manual' instead of 'heat' to system mode
            // Disabled as write failes (see above)
            // .withLocalTemperature(ea.STATE).withSystemMode(['off', 'auto', 'heat'], ea.STATE_SET),
        e.window_detection(),
        e.child_lock(),
        e.battery(),
        exposes.binary('window_open', ea.STATE).withDescription('Window open?'),
    ],
    meta: {
        tuyaDatapoints: [
            [2, 'current_heating_setpoint', tuya.valueConverter.divideBy10],
            [3, 'local_temperature', tuya.valueConverter.divideBy10],
            [4, 'system_mode', tuya.valueConverterBasic.lookup({'off': tuya.enum(0), 'auto': tuya.enum(1), 'heat': tuya.enum(2)})],
            [7, 'child_lock', tuya.valueConverterBasic.lookup({'LOCK': true, 'UNLOCK': false})],
            [17, 'window_open', tuya.valueConverterBasic.lookup({true: tuya.enum(1), false: tuya.enum(0)})],
            [18, 'window_detection', tuya.valueConverter.onOff],
            [21, 'battery', tuya.valueConverter.raw],
        ]
    },
};

module.exports = definition;
```

The windows_open is always false - BUT - at the moment I have 21°C in my Flat and there was 14-20°C outside during the day, maybe the code is wrong, the indicator not working OR it's simply to warm outside to let the valve recognize the temperature drop ... or maybe because the heating is off caused by warm weather and high gas prices ... anyway, I'll see whats the case in a few days or months.

Questions / Todo:
- Write system_mode possible or is it read-only?
- How to get 'manual' instead of 'heat' to system mode
- Whitelabel seems to have no effect?
- How to achive to get the model HTCZ01 and vendor Brennenstuhl and an icon (this may be the icon of the TS0601_thermostat) to the webinterface
- How to integrate that and contribute finally to the project

**Reply by csidirop** [reply]
My order arrived yesterday, so I can test a bit more.

> I maybe need to double check the data points, something looked odd. I try as soon as I can and give feedback.

Was everything right.

> Finally made some progress:

I merge your additions to my edits and got:
![grafik](https://user-images.githubusercontent.com/43964592/202284182-f17a939f-e4d8-44aa-a423-9ff2dcffc263.png)

I was testing the window detection and it dont work. I left the window open for an hour and the local temperature droped for around 15° in total. So there is something off.

**Reply by csidirop** [reply]
>     * How to get 'manual' instead of 'heat' to system mode

I can trigger a change in HA, but Z2M reports "No converter available for 'system_mode' ("off")"

The actual question is, what does the device expect or accept for requests? Is there any way to check what the original bridge is sending for requests?

**Reply by marcelebbrecht** [reply]
Setting this via MQTT is simply disabled as shown in my comments - you can enable this by setting

```
.withLocalTemperature(ea.STATE).withSystemMode(['off', 'auto', 'heat'], ea.STATE),
```

TO

```
.withLocalTemperature(ea.STATE).withSystemMode(['off', 'auto', 'heat'], ea.STATE_SET),
```

AND

```
toZigbee: [tzDataPoints],
```

TO

```
toZigbee: [tzDataPoints, tz.thermostat_system_mode],
```

But when you change the value, you get:
```
2022-11-01 19:57:38: Error: Write 0x84fd27fffe8af1b8/1 hvacThermostat({"systemMode":1}, {"sendWhen":"immediate","timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null,"writeUndiv":false}) failed (Status 'UNSUPPORTED_ATTRIBUTE')
at Endpoint.checkStatus (/srv/zigbee2mqtt/zigbee2mqtt/node_modules/zigbee-herdsman/src/controller/model/endpoint.ts:317:28)
at Endpoint.write (/srv/zigbee2mqtt/zigbee2mqtt/node_modules/zigbee-herdsman/src/controller/model/endpoint.ts:397:22)
at Object.convertSet (/srv/zigbee2mqtt/zigbee2mqtt/node_modules/zigbee-herdsman-converters/converters/toZigbee.js:1199:13)
at Publish.onMQTTMessage (/srv/zigbee2mqtt/zigbee2mqtt/lib/extension/publish.ts:246:36)
```

Maybe Koenkk can bring some light? But maybe, changing this value remote is simply unsupported by the device

**Reply by marcelebbrecht** [reply]
I updated the config above by adding child-lock and window-detection support for MQTT but system_mode is still not working

**Reply by csidirop** [reply]
>  But maybe, changing this value remote is simply unsupported by the device

It is, I tried it with the TuYa App. There are 3 modes:

**Auto:** (with timer activated)
![Auto](https://user-images.githubusercontent.com/43964592/204876135-ebdc4646-e81e-4e04-8c1c-fe4413a65324.jpg)

**Manual:**
![Manual](https://user-images.githubusercontent.com/43964592/204876252-37bef42e-3976-4c4a-ab69-6a7434a0044b.jpg)

**Freeze:** (Range: 5°-15°)
![Freeze](https://user-images.githubusercontent.com/43964592/204876316-bdb7072b-2001-4309-a626-2eab24fc65c3.jpg)

I took the mode titles from the TuYa IoT Platform, filtering the Logs. Is there any way to grab what ever the TuYa Gateway is sending?

**Reply by marcelebbrecht** [reply]
It is always good to know what won't work :) Anyway, I use OpenHAB and wrote rules to use the modes as indicator if the schedule should be overridden by manual settings.

**Comment by csidirop** [comment]
For the record: Currently the thermostat can be integrated [this way](https://github.com/Koenkk/zigbee2mqtt/discussions/14080#discussioncomment-3682723), but has only limited functionality:

- `child_lock` switch works
- `valve_detection` switch works
- `current_heating_setpoint` value works (in both directions)
- `local_temperature` works (quite accurate)
- _maybe more ..._

**Comment by marcelebbrecht** [comment]
Try Uppercase ?

Viele Grüße | Kind regards
Marcel Ebbrecht

PGP: http://gpg.fundm.eu/0x47EA9A32.asc

Am 17. November 2022 09:01:45 MEZ schrieb Chris ***@***.***>:

**Comment by peterhoeg** [comment]
Did anyone get any further supporting this? And @csidirop, I hope this isn't too off-topic, but any specific trick to get it to pair?

**Reply by csidirop** [reply]
Hi @peterhoeg , no, I didn't have time to do it anymore... But it was added some time ago and its working very smooth now. Thanks btw!
